### PR TITLE
KAFKA-6870 Concurrency conflicts in SampledStat

### PR DESCRIPTION
**Pull request by chia7712**
see [KAFKA-6870](https://issues.apache.org/jira/browse/KAFKA-6870) for more details.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by rajinisivaram** [comment]
@chia7712 Can you add one unit test please?

**Comment by chia7712** [comment]
> Can you add one unit test please?

Copy that.

File: clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java [hunk 1]
```diff
@@ -81,10 +79,13 @@ public Measurable measurable() {
...
-        if (this.metricValueProvider instanceof Measurable)
-            return ((Measurable) metricValueProvider).measure(config, timeMs);
-        else
-            return 0;
+        synchronized (this.lock) {
+            if (this.metricValueProvider instanceof Measurable)
+                return ((Measurable) metricValueProvider).measure(config, timeMs);
+            else
+                return 0;
+        }
```
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java:87]: nit: unnecessary newline

File: clients/src/test/java/org/apache/kafka/common/metrics/SensorTest.java [hunk 2]
```diff
@@ -124,4 +134,48 @@ public void testIdempotentAdd() {
...
+                        }
+                        return null;
+                    } catch (Throwable e) {
+                        return e;
+                    }
+                }
+            }));
+        }
+        latch.countDown();
+        service.shutdown();
```
Comment (rajinisivaram) [inline clients/src/test/java/org/apache/kafka/common/metrics/SensorTest.java]: we need to shutdown the executor service even if there is a test failure (either in a finally block or the teardown).

File: clients/src/test/java/org/apache/kafka/common/metrics/SensorTest.java [hunk 3]
```diff
@@ -124,4 +134,48 @@ public void testIdempotentAdd() {
...
+        assertTrue(service.awaitTermination(10, TimeUnit.SECONDS));
+        for (Future<Throwable> callable : workers) {
+            assertTrue(callable.isDone());
+            assertNull(callable.get());
```
Comment (rajinisivaram) [inline clients/src/test/java/org/apache/kafka/common/metrics/SensorTest.java]: Can we add error messages for these assertions?

**Review by rajinisivaram — COMMENTED** [review]
@chia7712 Thanks for the PR. Looks good, left a few minor comments.

**Comment by chia7712** [comment]
Thanks for the reviews. @rajinisivaram I have addressed you comment in the latest commit. Please take a look.

**Review by rajinisivaram — COMMENTED** [review]
@chia7712 Thanks for the update, LGTM. Will merge after PR builds complete.

**Review by rajinisivaram — APPROVED** [review]
@chia7712 Thanks for the PR, merging to trunk and 1.1.

**Pull request merged**